core/domain/imps_jobs_one_off.py

```diff
@@ -0,0 +1,205 @@
+# Copyright 2020 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""One off jobs related to Oppia improvement tasks."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import ast
+import collections
+
+from core.domain import prod_validation_jobs_one_off
+from core.platform import models
+import feconf
+import python_utils
+
+exp_models, imps_models = models.Registry.import_models([
+    models.NAMES.exploration, models.NAMES.improvements])
+
+
+def _target_exists(entity, target_type, target_id):
+    """Checks whether the given target exists in the entity."""
+    if isinstance(entity, exp_models.ExplorationModel):
+        # Returning with this style allows us to hit 100% code coverage.
+        if target_type == 'state' and target_id in entity.states:
+            return True
+    return False
+
+
+class TaskEntryModelAuditOneOffJob(
```

Comment:
Why can't we follow the procedure similar to other jobs in `prod_validation_jobs_one_off`? We first define a validator which has all the checks and the audit job uses the validator to validate the models. This will make the code more readable and consistent.

Reply:
Ah, I hadn't actually noticed that pattern when writing this. Taking a look now, there are two main issues I see:
1.  The reduce function doesn't seem accessible? These models also need to compare against "each other" to check that two tasks don't refer to the same version.
2.  These models need to validate external IDs _and_ version numbers, so only checking for IDs is insufficient.

Not sure how to address these points without refactoring the validator a lot?

Reply:
Regarding 1, I think it is fine to have a separate audit job only for the comparison as we have here as well (https://github.com/oppia/oppia/blob/develop/core/domain/prod_validation_jobs_one_off.py#L5668).

Regarding 2, I am sorry I did not understand the query, the external IDs validation is supported in the BaseModelValidator, so we can use that right?

Reply:
I've significantly cleaned up the logic, maybe my points are clearer now? PTAL!

Reply:
Thanks for cleaning up, that helps! Left a few comments, PTAL.